core/arch/arm/plat-zynqmp/sta_bsload.c

```diff
@@ -0,0 +1,366 @@
+#define CSU_SSS_CFG_PCAP_SSS_MASK	0xF
+#define CSU_SSS_CFG_PCAP_SSS_SHFT	0x0
+#define CSU_SSS_CFG_PCAP_DMA_DATA	0x5
+#define CSU_PCAP_RESET			(csu_base + 0x300C)
+#define CSU_PCAP_RDWR			(csu_base + 0x3004)
+#define CSU_PCAP_CTRL			(csu_base + 0x3008)
+#define CSU_PCAP_PROG			(csu_base + 0x3000)
+#define CSU_PCAP_STATUS			(csu_base + 0x3010)
+#define CSU_PCAP_RESET_RESET		BIT(0)
+#define CSU_PCAP_CTRL_PCAP_PR		BIT(0)
+#define CSU_PCAP_PROG_PCFG_PROG_B	BIT(0)
+#define CSU_PCAP_STATUS_PCAP_WR_IDLE	BIT(0)
+#define CSU_PCAP_STATUS_PL_INIT		BIT(2)
+#define CSU_PCAP_STATUS_PL_DONE		BIT(3)
+#define CSU_PCAP_STATUS_PL_RESET_B	BIT(6)
+
+#define CSUDMA_SRC_ADDR			(csudma_base + 0x000)
+#define CSUDMA_SRC_CRC			(csudma_base + 0x010)
+#define CSUDMA_SRC_SIZE			(csudma_base + 0x004)
+#define CSUDMA_SRC_I_STS		(csudma_base + 0x014)
+#define CSUDMA_SRC_ADDR_MSB		(csudma_base + 0x028)
+#define CSUDMA_SRC_I_STS_DONE		BIT(1)
+#define CSUDMA_SRC_SIZE_MASK		0x1ffffffc
+
+#define DUMMY_WORD			0xffffffff
+#define BS_SIZE_MAX			0x1fffffff
+
+/* Xilinx binary format header */
+static const uint32_t bin_format[] = {
+	DUMMY_WORD, /* Dummy words */
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	DUMMY_WORD,
+	0x000000bb, /* Sync word */
+	0x11220044, /* Sync word */
+	DUMMY_WORD,
+	DUMMY_WORD,
+	0xaa995566, /* Sync word */
+};
+
+static TEE_Result poll_reg_tout_32(vaddr_t addr, uint32_t mask, uint32_t val,
+				   ssize_t tout)
+{
+	uint32_t reg;
+
+	do {
+		reg = read32(addr);
+	} while (((reg & mask) != val) && tout--);
+
+	if (tout <= 0)
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result check_header(const paddr_t buf)
+{
+	const uint32_t *test = (const uint32_t *)buf;
+
+	for (size_t i = 0; i < ARRAY_SIZE(bin_format); i++) {
+		if (test[i] != bin_format[i])
+			return TEE_ERROR_BAD_FORMAT;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pcap_init(void)
+{
+	TEE_Result ret;
+
+	/* reset PCAP */
+	write32(CSU_PCAP_RESET_RESET, CSU_PCAP_RESET);
+	write32(0, CSU_PCAP_RESET);
+
+	write32(CSU_PCAP_CTRL_PCAP_PR, CSU_PCAP_CTRL);
+	write32(0, CSU_PCAP_RDWR);
+
+	/* reset PL */
+	write32(0, CSU_PCAP_PROG);
+	ret = poll_reg_tout_32(CSU_PCAP_STATUS, CSU_PCAP_STATUS_PL_RESET_B, 0,
+			       0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout waiting for PL reset\n");
+		return ret;
+	}
+
+	write32(CSU_PCAP_PROG_PCFG_PROG_B, CSU_PCAP_PROG);
+	ret = poll_reg_tout_32(CSU_PCAP_STATUS, CSU_PCAP_STATUS_PL_INIT,
+			       CSU_PCAP_STATUS_PL_INIT, 0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout waiting for PL init done\n");
+		return ret;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pcap_write(uint64_t srcaddr, uint32_t size)
+{
+	TEE_Result ret;
+	uint32_t regval;
+
+	write32(0, CSUDMA_SRC_CRC);
+
+	regval = read32(CSU_SSS_CONFIG);
+	regval &= CSU_SSS_CFG_PCAP_SSS_MASK;
```

Comment:
Just a quick note that io_mask* functions expect that val and mask to be already shifted as well.